tests/drivers/disk/src/main.c

```diff
@@ -0,0 +1,219 @@
+	} else {
+		zassert_not_equal(rc, 0, "Disk should fail to read out of sector bounds");
+	}
+}
+
+/* Write sector of disk, and check the data to ensure it is valid
+ * WARNING: this test is destructive- it will overwrite data on the disk!
+ */
+static int write_sector_checked(uint8_t *wbuf, uint8_t *rbuf,
+			uint32_t start, uint32_t num_sectors)
+{
+	int rc, i;
+
+	/* First, fill the write buffer with data */
+	for (i = 0; i < num_sectors * disk_sector_size; i++) {
+		wbuf[i] = (i & (~num_sectors));
+	}
+	/* Now write data to the sector */
+	rc = disk_access_write(disk_pdrv, wbuf, start, num_sectors);
+	if (rc) {
+		return rc; /* Let calling function handle disk error */
+	}
+	/* Read back the written data into another buffer */
+	memset(rbuf, 0, num_sectors * disk_sector_size);
+	rc = read_sector(rbuf, start, num_sectors);
+	if (rc) {
+		return rc;
+	}
+	/* Check the read data versus the written data */
+	zassert_equal(memcmp(wbuf, rbuf, num_sectors * disk_sector_size), 0,
+		"Read data did not match data written to disk");
+	return rc;
+}
+
+/* Tests writing to a variety of sectors
+ * WARNING: this test is destructive- it will overwrite data on the disk!
+ */
+static void test_sector_write(uint8_t *wbuf, uint8_t *rbuf, uint32_t num_sectors)
+{
+	int rc, sector;
+
+	TC_PRINT("Testing writes of %u sectors\n", num_sectors);
+	/* Write to disk sector zero */
+	rc = write_sector_checked(wbuf, rbuf, 0, num_sectors);
+	zassert_equal(rc, 0, "Failed to write to sector zero");
+	/* Write to a sector in the "middle" of the disk */
+	sector = MAX(((disk_sector_count / 2) - num_sectors), 0);
+	rc = write_sector_checked(wbuf, rbuf, sector, num_sectors);
+	zassert_equal(rc, 0, "Failed to write to mid disk sector");
+	/* Write to the last sector */
+	rc = write_sector_checked(wbuf, rbuf, disk_sector_count - 1, num_sectors);
+	if (num_sectors == 1) {
+		zassert_equal(rc, 0, "Failed to write to last sector");
+	} else {
+		zassert_not_equal(rc, 0, "Disk should fail to write out of sector bounds");
+	}
+}
+
+/* Test multiple reads in series, and reading from a variety of blocks */
+static void test_read(void)
+{
+	int rc, i;
+
+	/* Verify all 4 read sizes work */
+	test_sector_read(scratch_buf[0], SECTOR_COUNT1);
+	test_sector_read(scratch_buf[0], SECTOR_COUNT2);
+	test_sector_read(scratch_buf[0], SECTOR_COUNT3);
+	test_sector_read(scratch_buf[0], SECTOR_COUNT4);
+
+	/* Verify that reading from the same location returns to same data */
+	memset(scratch_buf[0], 0, SECTOR_COUNT1 * disk_sector_size);
+	rc = read_sector(scratch_buf[0], 0, SECTOR_COUNT1);
+	zassert_equal(rc, 0, "Failed to read from disk");
+	i = 10;
+	while (i--) {
+		/* Read from sector, and compare it to the first read */
+		memset(scratch_buf[1], 0xff, SECTOR_COUNT1 * disk_sector_size);
+		rc = read_sector(scratch_buf[1], 0, SECTOR_COUNT1);
+		zassert_equal(rc, 0, "Failed to read from disk at same sector location");
+		zassert_equal(memcmp(scratch_buf[1], scratch_buf[0],
```

Comment:
```diff
diff --git a/tests/drivers/disk/src/main.c b/tests/drivers/disk/src/main.c
index b276b649085..c76af9cde5d 100644
--- a/tests/drivers/disk/src/main.c
+++ b/tests/drivers/disk/src/main.c
@@ -178,9 +178,9 @@ static void test_read(void)
                memset(scratch_buf[1], 0xff, SECTOR_COUNT1 * disk_sector_size);
                rc = read_sector(scratch_buf[1], 0, SECTOR_COUNT1);
                zassert_equal(rc, 0, "Failed to read from disk at same sector location");
-               zassert_equal(memcmp(scratch_buf[1], scratch_buf[0],
-                               SECTOR_COUNT1 * disk_sector_size), 0,
-                               "Multiple reads from same disk location did not return same data");
+               zassert_mem_equal(scratch_buf[1], scratch_buf[0],
+                                 SECTOR_COUNT1 * disk_sector_size,
+                                 "Multiple reads mismatch");
        }
 }
```

Please keep it short.

Reply:
you can use `zassert_mem_equal` instead of zassert_equal(memcmp())

```
+               zassert_mem_equal(scratch_buf[1], scratch_buf[0],
+                                 SECTOR_COUNT1 * disk_sector_size,
+                                 "Multiple reads mismatch");
```

Reply:
fixed, sorry I missed that change before.